examples/pytorch/graphsaint/train_sampling.py

```diff
@@ -1,128 +1,87 @@
     data = load_data(args, multilabel)
     g = data.g
-    train_mask = g.ndata['train_mask']
-    val_mask = g.ndata['val_mask']
-    test_mask = g.ndata['test_mask']
-    labels = g.ndata['label']
-
-    train_nid = data.train_nid
-
-    in_feats = g.ndata['feat'].shape[1]
-    n_classes = data.num_classes
-    n_nodes = g.num_nodes()
-    n_edges = g.num_edges()
-
-    n_train_samples = train_mask.int().sum().item()
-    n_val_samples = val_mask.int().sum().item()
-    n_test_samples = test_mask.int().sum().item()
-
-    print("""----Data statistics------'
-    #Nodes %d
-    #Edges %d
-    #Classes/Labels (multi binary labels) %d
-    #Train samples %d
-    #Val samples %d
-    #Test samples %d""" %
-          (n_nodes, n_edges, n_classes,
-           n_train_samples,
-           n_val_samples,
-           n_test_samples))
-    # load sampler
-
-    kwargs = {
-        'dn': args.dataset, 'g': g, 'train_nid': train_nid, 'num_workers_sampler': args.num_workers_sampler,
-        'num_subg_sampler': args.num_subg_sampler, 'batch_size_sampler': args.batch_size_sampler,
-        'online': args.online, 'num_subg': args.num_subg, 'full': args.full
-    }
-
-    if args.sampler == "node":
-        saint_sampler = SAINTNodeSampler(args.node_budget, **kwargs)
-    elif args.sampler == "edge":
-        saint_sampler = SAINTEdgeSampler(args.edge_budget, **kwargs)
-    elif args.sampler == "rw":
-        saint_sampler = SAINTRandomWalkSampler(args.num_roots, args.length, **kwargs)
-    else:
-        raise NotImplementedError
-    loader = DataLoader(saint_sampler, collate_fn=saint_sampler.__collate_fn__, batch_size=1,
-                        shuffle=True, num_workers=args.num_workers, drop_last=False)
+    train_g = dgl.node_subgraph(g, data.train_nid)
+    train_g.ndata['D_norm'] = 1. / train_g.in_degrees().clamp(min=1).unsqueeze(1)
+    g.ndata['D_norm'] = 1. / g.in_degrees().clamp(min=1).unsqueeze(1)
+
+    print('Pre-computing normalization coefficients')
+    sampler = SAINTSampler(mode=args.sampler, budget=args.budget)
+    loader = DataLoader(train_g, torch.arange(1000), sampler, num_workers=args.num_workers)
+
+    num_train_nodes = train_g.num_nodes()
+    node_count = torch.zeros(num_train_nodes)
+    edge_count = torch.zeros(train_g.num_edges())
+
+    target_num_nodes = num_train_nodes * args.norm_ratio
+    num_sgs = 0
+    num_sampled_nodes = 0
+    pbar = tqdm(total=target_num_nodes)
+
+    while num_sampled_nodes < target_num_nodes:
+        for sg in loader:
+            node_count[sg.ndata[dgl.NID]] += 1
+            edge_count[sg.edata[dgl.EID]] += 1
+            num_sampled_nodes += sg.num_nodes()
+            pbar.update(sg.num_nodes())
+            num_sgs += 1
+    pbar.close()
+
+    train_g.ndata['n_c'] = node_count.clamp(min=1)
+    train_g.edata['e_c'] = edge_count.clamp(min=1)
+    train_g.apply_edges(fn.v_div_e('n_c', 'e_c', 'w'))
+    train_g.ndata['l_n'] = num_sgs / train_g.ndata['n_c'] / num_train_nodes
```

Comment:
I feel all the above should be part of `SAINTSampler`'s preprocessing and should not be exposed to users, like `ClusterGCNSampler`.

Reply:
This is something I'm not sure based on previous discussion with @jermainewang . To speedup the computation of it we might want to use multiple processes, which might make the sampler too complex. Also this depends on whether in most use cases users are only interested in sampling subgraphs without the normalization coefficients.

Reply:
I was too late.  This could be done by having `SAINTSampler` with two modes, one that computes the norm and one that does not (like you already did).  The norm could be computed in a multiprocessing way with a DataLoader together with `SAINTSampler` not computing the norm.  I think this is how PyG does.

Reply:
Yes, this is what I meant. We will need to construct a `DataLoader` inside `SAINTSampler` to sample a number of subgraphs with a nonzero `num_workers` and use these subgraphs to compute normalization coefficients and optionally cache them. PyG's implementation allows optionally doing that with `num_workers=0`. This was my initial design and Minjie found it too complex with three norm-specific arguments.